ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,139 @@
+(*
+ * Copyright (c) Cloud Software Group, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Helpers
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+open D
+
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let inode_size = 4096L
+
+let ( ++ ) = Int64.add
+
+let ( // ) = Filename.concat
+
+exception Illegal_file_path of string
+
+exception Unsupported_file_type of string
+
+exception File_exceeds_max_size_limit of string
+
+exception File_incomplete of string
+
+exception File_corrupted of string
+
+exception File_parsed_failed of string
+
+type result = Success | Failure of exn
+
+type unpack_file_result = Next of int64 | Error of exn | Done
```

Comment:
(May not be stated clearly in my previous comment)  
What I meant to is something like:
```
type unpack_error =
  | Illegal_file_path of string
  | Unsupported_file_type of string
  | Unpacked_exceeds_max_size_limit of Int64.t
  | File_size_mismatch of (Int64.t * Int64.t)
  | File_incomplete
  | File_corrupted
  | Unpacking_failure
```
And then the `unpack_tar_file` could be:
```
let unpack_tar_file ~dir ~ifd ~max_size_limit =
  debug "Unpacked to %s" dir ;
  Unixext.rm_rec dir ;
  Unixext.mkdir_rec dir dir_perm ;
  let rec unpack_all_files unpack dir acc_size max_size_limit =
    match
      Tar_unix.Archive.with_next_file ifd (unpack dir acc_size max_size_limit)
    with
    | exception Tar.Header.End_of_stream ->
        Ok ()
    | Ok (Next size) ->
        unpack_all_files unpack dir size max_size_limit
    | Error _ as err ->
        err
    | exception End_of_file ->
        Error File_incomplete
    | exception Tar.Header.Checksum_mismatch ->
        Error File_corrupted
    | exception ex ->
        error "Failed to unpack to %s with exception: %s" dir
          (Printexc.to_string ex) ;
        Error Unpacking_failure
  in
  unpack_all_files unpack dir 0L max_size_limit
  |> Result.map_error (fun err -> Unixext.rm_rec dir ; err)
```